### DISCUSSION TITLE
kindling-collector运行占用cpu过高

### DISCUSSION THREAD

**Discussion by zcl1115** [harmonycloud/kindling#597] [agent collector]
如题，kindling-collector在部分机器上运行，cpu占用过高，影响到了业务，想问一下有什么优化解决方案吗，已经加了2C的cgroup限制了

**Comment by dxsup** [comment]
能否提供一下环境信息？
- 主机Pod数
- CPU核心数
- CPU使用率
- kindling-collector的CPU使用率

**Reply by zcl1115** [reply]
非k8s环境，探测物理机数据
cpu核心数：2*10
cpu使用率：
![image](https://github.com/KindlingProject/kindling/assets/45005291/7f504071-0129-453d-b41b-cb422d2d72fe)
kindling-collector的CPU使用率：
![image](https://github.com/KindlingProject/kindling/assets/45005291/3df53d6b-2ab7-4e8f-8bf3-05c72fc4261f)

**Reply by dxsup** [reply]
能否解释一下CPU使用率的变化是什么情况下发生的？是开启kindling-collector后CPU上升到100%，然后关闭探针后恢复到60%吗？

**Reply by zcl1115** [reply]
是的，机器内核版本是
Linux bigo 4.9.70-040970-generic #201712161132 SMP Sat Dec 16 16:33:52 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
如下图所示，在部署kindling-collector到卸载（上午跟刚才一阵）期间，机器上所有的服务的占用cpu都有所上涨，卸载探针后全部下降恢复到正常水平
![image](https://github.com/KindlingProject/kindling/assets/45005291/a96697fa-0a12-49cd-9906-3e9475df6073)

**Reply by dxsup** [reply]
根据我们的测试，启动kindling-collector几乎不会对进程本身造成影响。
- 业务是用什么编程语言编写的？
- 是否开启了Trace Profiling功能？
- 能否用火焰图工具在启动前和启动后分别生成一次火焰图，看是哪部分代码占用CPU上升了？

**Reply by zcl1115** [reply]
业务是brpc的C++语言编写的，没有开启trace profiling功能，我们部署的版本是基于开源稳定分支的基础上加了一个processor，打上了机器采集的进程名称相关信息，kindling-collector本身启动后的火焰图如下，我这边分析不出来哪里有上升，但是看到一个很深的调用栈
![Uploading kindling_new.svg…]()

**Reply by zcl1115** [reply]
https://github.com/zcl1115/kindling_svg

**Reply by zcl1115** [reply]
因为机器内核版本过低，4.9.70，所以这里没有使用load_ebpf_probe，而是采用的load_kernel_probe，不知道这个会不会也有影响

**Reply by dxsup** [reply]
kindling-collector的火焰图大概了解了，使用load_kernel_probe一般性能会更好。
目前奇怪的是业务进程的CPU会上升，这是我们之前没有遇到的。你可以看下业务的火焰图，看业务是只有部分代码在火焰图上的占比增加了，还是业务所有代码的耗时都有升高而每个部分占比不变。

**Reply by zcl1115** [reply]
这里面的old.svg跟new.svg是我在collector启动的前后抓取一个测试进程的火焰图变化https://github.com/zcl1115/kindling_svg   ，看起来是占比不变而是整体耗时增加

**Reply by dxsup** [reply]
麻烦在配置文件的`receivers.cgoreceiver.subscribe`中，删除下面四个`kprobe`再重启一下kindling-collector，观察一下业务进程的CPU变化。
```yaml
receivers:
  cgoreceiver:
    subscribe:
      - name: kprobe-tcp_close
      - name: kprobe-tcp_rcv_established
      - name: kprobe-tcp_drop
      - name: kprobe-tcp_retransmit_skb
```

**Reply by zcl1115** [reply]
还是在上涨，现在这个配置是这样的
receivers:
  cgoreceiver:
    subscribe:
      - name: syscall_exit-writev
        category: net
      - name: syscall_exit-readv
        category: net
      - name: syscall_exit-write
        category: net
      - name: syscall_exit-read
        category: net
      - name: syscall_exit-sendto
        category: net
      - name: syscall_exit-recvfrom
        category: net
      - name: syscall_exit-sendmsg
        category: net
      - name: syscall_exit-recvmsg
        category: net
      - name: syscall_exit-sendmmsg
        category: net
      - name: syscall_exit-connect
      - name: kretprobe-tcp_connect
      - name: kprobe-tcp_set_state
      - name: tracepoint-procexit
    process_filter:
      # the length of a comm should be no more than 16
      comms:
        - "kindling-collec"
        - "containerd"
        - "dockerd"
        - "containerd-shim"

**Reply by dxsup** [reply]
```yaml
- name: kretprobe-tcp_connect
- name: kprobe-tcp_set_state
```
这两个也删掉。会涨的小一些吗？

**Reply by zcl1115** [reply]
刚才重启有个业务进程cpu暴涨到250左右，我这边perf抓取了下火焰图在这里  https://github.com/zcl1115/kindling_svg/blob/main/hight_cpu.svg

**Reply by zcl1115** [reply]
还有一个现象是我这边停止了卸载了kindling-collector，同时rmmod kindling_falcolib_probe，但是观察到cpu的系统态时间一直降不下来
![image](https://github.com/KindlingProject/kindling/assets/45005291/4514e785-a7da-4c3a-ae13-0d418bc5c165)

**Reply by dxsup** [reply]
有没有在高版本内核上运行过kindling-collector？是否有类似的问题？

**Reply by zcl1115** [reply]
有，但是没有遇到过这种问题，同时我在别的机器上有同样进程的机器部署了，cpu总使用率也没有很明显的涨幅（30->40），那个进程不是我们这边监控团队的服务，是业务团队的，我们拿不到权限，之前部署我们内部自己的机器以及上午部署90台测试机器，别的都没有这种情况

**Reply by dxsup** [reply]
这台机器与那些机器的区别就在于内核版本吗？是否还存在其他变量？麻烦再看一下内核日志，看有没有kindling相关的信息。

**Reply by zcl1115** [reply]
有类似这种内核日志，跟别的机器不太一样
Nov 27 10:35:55 bigo kernel: : [17424363.004500] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 10:35:55 bigo kernel: : [17424391.014680] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 10:35:55 bigo kernel: : [17424418.949634] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 10:35:55 bigo kernel: : [17424446.952238] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 10:35:55 bigo kernel: : [17424474.947042] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 10:35:55 bigo kernel: : [17424502.959421] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 14:23:22 bigo kernel: : [17438010.706391] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul
Nov 27 14:23:22 bigo kernel: : [17438038.676255] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) sctp libcrc32c dccp_diag dccp unix_diag af_packet_diag netlink_diag nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables xt_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag nvme nvme_core ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv4 nf_defrag_ipv4 nf_conntrack 8021q garp mrp stp llc bonding binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp kvm_intel kvm irqbypass intel_rapl_perf input_leds joydev ioatdma mei_me dca lpc_ich mei shpchp nfit mac_hid acpi_power_meter sch_fq tcp_bbr autofs4 btrfs xor raid6_pq hid_generic usbhid hid crct10dif_pclmul crc32_pclmul

**Reply by zcl1115** [reply]
Nov 27 10:57:23 bigo kernel: : [9332592.513335] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) dccp_diag dccp unix_diag af_packet_diag netlink_diag ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables nf_conntrack_ipv4 nf_defrag_ipv4 xt_conntrack nf_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp 8021q garp mrp stp llc kvm_intel kvm irqbypass intel_rapl_perf joydev input_leds lpc_ich mei_me mei ioatdma shpchp dca nfit acpi_power_meter mac_hid sch_fq tcp_bbr bonding autofs4 btrfs xor raid6_pq crct10dif_pclmul crc32_pclmul ghash_clmulni_intel hid_generic aesni_intel aes_x86_64
Nov 27 10:57:23 bigo kernel: : [9332592.513402]  [<ffffffffc08dc6a7>] record_event_consumer.part.9+0x817/0xda0 [kindling_falcolib_probe]
Nov 27 10:57:23 bigo kernel: : [9332592.513413]  [<ffffffffc08dcc93>] record_event_all_consumers+0x63/0x90 [kindling_falcolib_probe]
Nov 27 10:57:23 bigo kernel: : [9332592.513416]  [<ffffffffc08dd29e>] syscall_exit_probe+0x10e/0x140 [kindling_falcolib_probe]
Nov 27 10:57:23 bigo kernel: : [9332620.485744] Modules linked in: nf_conntrack_netlink kindling_falcolib_probe(OE) dccp_diag dccp unix_diag af_packet_diag netlink_diag ipmi_si ipmi_devintf ipmi_watchdog ipmi_msghandler bridge nf_conntrack_ipv6 nf_defrag_ipv6 ip6table_filter ip6_tables nf_conntrack_ipv4 nf_defrag_ipv4 xt_conntrack nf_conntrack xt_multiport nf_log_ipv4 nf_log_common xt_LOG xt_set ip_set_hash_net ip_set nfnetlink iptable_filter ip_tables x_tables tcp_diag udp_diag inet_diag binfmt_misc intel_rapl skx_edac edac_core x86_pkg_temp_thermal coretemp 8021q garp mrp stp llc kvm_intel kvm irqbypass intel_rapl_perf joydev input_leds lpc_ich mei_me mei ioatdma shpchp dca nfit acpi_power_meter mac_hid sch_fq tcp_bbr bonding autofs4 btrfs xor raid6_pq crct10dif_pclmul crc32_pclmul ghash_clmulni_intel hid_generic aesni_intel aes_x86_64
Nov 27 10:57:23 bigo kernel: : [9332620.485820]  [<ffffffffc08dc6a7>] record_event_consumer.part.9+0x817/0xda0 [kindling_falcolib_probe]
Nov 27 10:57:23 bigo kernel: : [9332620.485834]  [<ffffffffc08dcc93>] record_event_all_consumers+0x63/0x90 [kindling_falcolib_probe]
Nov 27 10:57:23 bigo kernel: : [9332620.485837]  [<ffffffffc08dd29e>] syscall_exit_probe+0x10e/0x140 [kindling_falcolib_probe]

**Reply by dxsup** [reply]
目前看起来是内核模块导致的异常，brpc的那个测试进程的源码能否开源出来？

**Reply by zcl1115** [reply]
那个测试进程不归属我们监控团队这边负责，是业务团队的，我们开源不了这个，而且感觉这个是跟机器本身有关的，不关乎业务进程的问题，我明天尝试在这台机器上再重新生成一个模块看看是否有异常

**Comment by dxsup** [comment]
brpc的那个测试进程的源码能否开源出来？我们测试看看。
